components/appmgt/org.wso2.carbon.appmgt.impl/src/main/java/org/wso2/carbon/appmgt/impl/dao/AppMDAO.java

```diff
@@ -1206,6 +1206,108 @@ public void removeSubscription(APIIdentifier identifier, int applicationId)
+			throws AppManagementException {
+
+		Connection conn = null;
+		ResultSet results = null;
+		PreparedStatement ps = null;
+		int fromAppId = -1;
+		int toAppId = -1;
+		int count;
+
+		try {
+
+			String getAppIdQuery = "SELECT APP_ID FROM APM_APP " +
+					"WHERE APP_PROVIDER=? AND APP_NAME=? AND APP_VERSION=?";
+
+			try {
+				conn = APIMgtDBUtil.getConnection();
+				ps = conn.prepareStatement(getAppIdQuery);
+			} catch (SQLException e) {
+				handleException("Failed to create database connection ", e);
+			}
+
+			try {
+				ps.setString(1, fromIdentifier.getProviderName());
+				ps.setString(2, fromIdentifier.getApiName());
+				ps.setString(3, fromIdentifier.getVersion());
+				results = ps.executeQuery();
+				if (results.next()) {
+					fromAppId = results.getInt(1);
+				}
+			} catch (SQLException e) {
+				APIMgtDBUtil.closeAllConnections(ps, conn, results);
+				String msg = "Could not retrieve app ID of " + fromIdentifier.getProviderName() +
+						"-" + fromIdentifier.getApiName() + "-" + fromIdentifier.getVersion();
+				handleException(msg, e);
+			}
+			APIMgtDBUtil.closeAllConnections(null, null, results);
+			if (fromAppId == -1) {
+				if (log.isDebugEnabled()) {
+					log.debug("Could not find app ID of 'from' app " + fromIdentifier);
+				}
+				return -1;
+			}
+
+			try {
+				ps.setString(1, toIdentifier.getProviderName());
+				ps.setString(2, toIdentifier.getApiName());
+				ps.setString(3, toIdentifier.getVersion());
+				results = ps.executeQuery();
+				if (results.next()) {
+					toAppId = results.getInt(1);
+				}
+			} catch (SQLException e) {
+				APIMgtDBUtil.closeAllConnections(ps, conn, results);
+				String msg = "Could not retrieve app ID of " + toIdentifier.getProviderName() + "-"
+						+ toIdentifier.getApiName() + "-" + toIdentifier.getVersion();
+				handleException(msg, e);
+			}
+			APIMgtDBUtil.closeAllConnections(null, null, results);
+			if (toAppId == -1) {
+				if (log.isDebugEnabled()) {
+					log.debug("Could not find app ID of 'to' app " + toIdentifier);
+				}
+				return -1;
+			}
+
+			String moveQuery = "UPDATE APM_SUBSCRIPTION SET APP_ID=? WHERE APP_ID=?";
+			count = -1;
+			try {
+				ps = conn.prepareStatement(moveQuery);
+				ps.setInt(1, toAppId);
+				ps.setInt(2, fromAppId);
+				count = ps.executeUpdate();
+				conn.commit();
+			} catch (SQLException e) {
+				String msg = "Could not move subscriptions from " +
+						fromIdentifier.getProviderName() + "-" + fromIdentifier.getApiName() +
+						"-" + fromIdentifier.getVersion() + " app to " +
+						toIdentifier.getProviderName() + "-" + toIdentifier.getApiName() +
+						"-" + toIdentifier.getVersion() + " app";
+				handleException(msg, e);
+			}
+
+		} finally {
+			APIMgtDBUtil.closeAllConnections(ps, conn, results);
+		}
+
+		if (log.isDebugEnabled()) {
+			log.debug(count + " subscribers were moved from" + fromIdentifier + " to " +
+							  toIdentifier);
```

Comment:
In those lines, a detailed message was need because of exception throwing.
However, now the `moveSubscriptions` method is broken into to methods, see commit 07d4afd84e007233c5a9ee791d8288c5bffb3cbf of PR #153